rmw_fastrtps_cpp/src/functions.cpp

```diff
@@ -1096,13 +1172,32 @@ rmw_subscription_t * rmw_create_subscription(const rmw_node_t * node,
     _register_type(participant, info->type_support_, info->typesupport_identifier_);
   }
 
-  SubscriberAttributes subscriberParam;
+  subscriberParam.historyMemoryPolicy = PREALLOCATED_WITH_REALLOC_MEMORY_MODE;
   subscriberParam.topic.topicKind = NO_KEY;
   subscriberParam.topic.topicDataType = type_name;
-  subscriberParam.topic.topicName = topic_name;
-  subscriberParam.historyMemoryPolicy = PREALLOCATED_WITH_REALLOC_MEMORY_MODE;
+  // set topic and partitions
+  name_tokens = rcutils_split_last(topic_name, '/');
+  if (name_tokens.size == 1) {
+    subscriberParam.qos.m_partition.push_back(ros_topics_prefix);
+    subscriberParam.topic.topicName = name_tokens.data[0];
+  } else if (name_tokens.size == 2) {
+    subscriberParam.qos.m_partition.push_back(
+      (std::string(ros_topics_prefix) + "/" + name_tokens.data[0]).c_str());
+    subscriberParam.topic.topicName = name_tokens.data[1];
+  } else {
+    RMW_SET_ERROR_MSG("Illformated topic name");
+    goto fail;
+  }
+  if (rcutils_string_array_fini(&name_tokens) != RCUTILS_RET_OK) {
+    fprintf(stderr, "Failed to destroy the token string array\n");
+  }
```

Comment:
This whole block seems to be shared between create publisher and create subscriber. Perhaps you should make a shared function which could be improved and tested in one place.